mm2src/coins/eth.rs

```diff
@@ -2670,150 +2683,158 @@ impl EthCoin {
         secret_hash: &[u8],
         amount: BigDecimal,
         expected_swap_contract_address: Address,
-    ) -> Box<dyn Future<Item = (), Error = String> + Send> {
-        let unsigned: UnverifiedTransaction = try_fus!(rlp::decode(payment_tx));
-        let tx = try_fus!(SignedEthTx::new(unsigned));
-        let sender = try_fus!(addr_from_raw_pubkey(sender_pub));
-        let expected_value = try_fus!(wei_from_big_decimal(&amount, self.decimals));
+    ) -> ValidatePaymentFut<()> {
+        let unsigned: UnverifiedTransaction = try_mm_err_fus!(rlp::decode(payment_tx));
+        let tx = try_mm_err_fus!(SignedEthTx::new(unsigned));
+        let sender =
+            try_mm_err_fus!(addr_from_raw_pubkey(sender_pub).map_to_mm(ValidatePaymentError::AddressParseError));
+        let expected_value = try_mm_err_fus!(wei_from_big_decimal(&amount, self.decimals));
         let selfi = self.clone();
         let secret_hash = secret_hash.to_vec();
         let fut = async move {
             let swap_id = selfi.etomic_swap_id(time_lock, &secret_hash);
-            let status = try_s!(
-                selfi
-                    .payment_status(expected_swap_contract_address, Token::FixedBytes(swap_id.clone()))
-                    .compat()
-                    .await
-            );
+            let status = selfi
+                .payment_status(expected_swap_contract_address, Token::FixedBytes(swap_id.clone()))
+                .compat()
+                .await
+                .map_to_mm(ValidatePaymentError::PaymentStatusError)?;
```

Comment:
`payment_status` function uses an RPC to retrieve the payment status, so we can consider the error as `TransportError`.